### PR TITLE
Support HTML markup for knob labels

### PR DESCRIPTION
**Pull request by liushuyu**
Add HTML rich text support for knob label

---
This change is [Reviewable](https://reviewable.io/reviews/lmms/lmms/3134)

### CODE REVIEW

**Comment by jasp00** [comment]
This is another example of why #3071 should be solved. Most of this patch is about format changes.

Regarding the main purpose, what kind of HTML effects do you plan to use that cannot be done through the style sheet? Is this feature for knob labels only or will this be extended to all labels?

**Comment by liushuyu** [comment]
>This is another example of why #3071 should be solved. Most of this patch is about format changes.

Right... That's because I have integrated a code formatter in `git` commit hook. So each time I commit something, the changed files will be formatted.

>Regarding the main purpose, what kind of HTML effects do you plan to use that cannot be done through the style sheet? Is this feature for knob labels only or will this be extended to all labels?

Well, that's relating to some chatter in our chat room: multi-line labels are not supported in knob labels, either `<br>` tag or `\n` didn't work at all.

**Comment by jasp00** [comment]
>multi-line labels are not supported in knob labels

Have you tried using QLabel?

**Comment by liushuyu** [comment]
>Have you tried using QLabel?  

QLabel works though.

**Comment by tresf** [comment]
@liushuyu what is our decision on this?  We can support HTML labels if it helps things, but I have to agree with @jasp00 about the unnecessary code formatting.

**Comment by Sawuare** [comment]
Bumping. ^

**Comment by PhysSong** [comment]
I reverted some formatting changes and resolved conflicts. There are some things to change more, e.g., fix a memory from `m_tdRenderer` by either setting `QObject` ownership or using `std::unique_ptr`, replacing `<font>` with `style` attribute.

File: src/gui/widgets/Knob.cpp [hunk 1]
```diff
@@ -172,6 +173,27 @@ void Knob::setLabel( const QString & txt )
...
+
+void Knob::setHtmlLabel(const QString &htmltxt)
+{
+	m_label = htmltxt;
+	m_isHtmlLabel = true;
+	// Put the rendered HTML content into cache
+	m_tdRenderer = new QTextDocument();
+
+	// TODO: stop using <font>
+	m_tdRenderer->setHtml(QString("<font color=\"%1\">%2</font>").arg(textColor().name(), m_label));
```
Comment (PhysSong) [inline src/gui/widgets/Knob.cpp:189]: Maybe something like this?
```suggestion
	m_tdRenderer->setHtml(QString("<span style=\"color:\"%1\;">%2</span>").arg(textColor().name(), m_label));
```

File: src/gui/widgets/Knob.cpp [hunk 2]
```diff
@@ -172,6 +173,27 @@ void Knob::setLabel( const QString & txt )
...
+	update();
+}
+
+
+void Knob::setHtmlLabel(const QString &htmltxt)
+{
+	m_label = htmltxt;
+	m_isHtmlLabel = true;
+	// Put the rendered HTML content into cache
+	m_tdRenderer = new QTextDocument();
```
Comment (PhysSong) [inline src/gui/widgets/Knob.cpp:186]: To fix memory leak:
```suggestion
	if (!m_tdRenderer)
	{
		m_tdRenderer = new QTextDocument(this);
	}
```

File: src/gui/widgets/Knob.cpp [hunk 3]
```diff
@@ -641,15 +663,20 @@ void Knob::paintEvent( QPaintEvent * _me )
...
-		p.setFont( pointSizeF( p.font(), 6.5 ) );
-/*		p.setPen( QColor( 64, 64, 64 ) );
-		p.drawText( width() / 2 -
-			p.fontMetrics().width( m_label ) / 2 + 1,
-				height() - 1, m_label );*/
-		p.setPen( textColor() );
-		p.drawText(width() / 2 -
+		if (!m_isHtmlLabel)
+		{
+			p.setFont(pointSizeF(p.font(),6.5));
```
Comment (PhysSong) [inline src/gui/widgets/Knob.cpp:668]:
```suggestion
			p.setFont(pointSizeF(p.font(), 6.5));
```

File: src/gui/widgets/Knob.cpp [hunk 4]
```diff
@@ -172,6 +173,29 @@ void Knob::setLabel( const QString & txt )
...
+{
+	m_label = htmltxt;
+	m_isHtmlLabel = true;
+	// Put the rendered HTML content into cache
+	if (!m_tdRenderer)
+	{
+		m_tdRenderer = new QTextDocument(this);
+	}
+
+	m_tdRenderer->setHtml(QString("<span style=\"color:\"%1\;">%2</span>").arg(textColor().name(), m_label));
```
Comment (PhysSong) [inline src/gui/widgets/Knob.cpp:191]:
```suggestion
	m_tdRenderer->setHtml(QString("<span style=\"color:\"%1;\">%2</span>").arg(textColor().name(), m_label));
```

File: src/gui/widgets/Knob.cpp [hunk 5]
```diff
@@ -172,6 +173,29 @@ void Knob::setLabel( const QString & txt )
...
+{
+	m_label = htmltxt;
+	m_isHtmlLabel = true;
+	// Put the rendered HTML content into cache
+	if (!m_tdRenderer)
+	{
+		m_tdRenderer = new QTextDocument(this);
+	}
+
+	m_tdRenderer->setHtml(QString("<span style=\"color:\"%1;\">%2</span>").arg(textColor().name(), m_label));
```
Comment (PhysSong) [inline src/gui/widgets/Knob.cpp:191]:
```suggestion
	m_tdRenderer->setHtml(QString("<span style=\"color:%1;\">%2</span>").arg(textColor().name(), m_label));
```

**Comment by PhysSong** [comment]
It's now ready to review/merge.

File: src/gui/widgets/Knob.cpp [hunk 6]
```diff
@@ -172,6 +174,29 @@ void Knob::setLabel( const QString & txt )
...
 	}
+
+	update();
+}
+
+
+void Knob::setHtmlLabel(const QString &htmltxt)
+{
+	m_label = htmltxt;
+	m_isHtmlLabel = true;
```
Comment (DomClark) [inline src/gui/widgets/Knob.cpp:185]: Should `setLabel` set `m_isHtmlLabel = false;` too? That way the label can be changed freely.

File: src/gui/widgets/Knob.cpp [hunk 7]
```diff
@@ -641,15 +666,20 @@ void Knob::paintEvent( QPaintEvent * _me )
...
+			p.setFont(pointSizeF(p.font(), 6.5));
+			p.setPen(textColor());
+			p.drawText(width() / 2 -
 				horizontalAdvance(p.fontMetrics(), m_label) / 2,
 				height() - 2, m_label);
+		}
+		else
+		{
+			m_tdRenderer->setDefaultFont(pointSizeF(p.font(), 6.5));
+			p.translate(width() / 2 - m_tdRenderer->idealWidth() / 2, (height() - m_tdRenderer->pageSize().height()) / 2);
```
[the remark this answers is not in the document]
Comment (PhysSong) [reply src/gui/widgets/Knob.cpp:680]: Good find, it does matter because the first division is a integer division.

**Review by DomClark — APPROVED** [review]

**Pull request merged**